### PR TITLE
add hardlink and arbitrary-options options

### PR DESCRIPTION
**Pull request by kwaegema**

### CODE REVIEW

File: bin/zkrsync.py [hunk 1]
```diff
@@ -246,7 +249,10 @@ def main():
         'verbose'     : ('run rsync with --verbose', None, 'store_true', False),
         # Individual Destination client specific options
         'rsyncport'   : ('force port on which rsyncd binds', "int", 'store', None),
-        'startport'   : ('offset to look for rsyncd ports', "int", 'store', 4444)
+        'startport'   : ('offset to look for rsyncd ports', "int", 'store', 4444),
+        # Arbitrary rsync options: comma seperate list. Use a colon to seperate key and values
+        'arbitopts'   : ('Arbitrary rsync source client options. Be careful', 'strlist', 'store', None),
```
Comment (stdweird) [inline bin/zkrsync.py]: add the comment to the help instead of a comment

File: lib/vsc/zk/rsync/source.py [hunk 2]
```diff
@@ -312,10 +314,15 @@ def get_flags(self, file, recursive):
...
             flags.append('--checksum')
+        if self.rsync_hardlinks:
+            flags.append('--hard-links')
         if self.rsync_verbose:
             flags.append('--verbose')
         if self.rsync_dry:
             flags.append('-n')
+        if self.rsync_arbitopts:
+            for arbopt in self.rsync_arbitopts:
+                flags.append("--%s" % re.sub(':', '=', arbopt))
```
Comment (stdweird) [inline lib/vsc/zk/rsync/source.py]: only replace the first occurrence of a `:`

File: lib/vsc/zk/rsync/source.py [hunk 3]
```diff
@@ -312,10 +314,15 @@ def get_flags(self, file, recursive):
...
         if self.rsync_checksum:
             flags.append('--checksum')
+        if self.rsync_hardlinks:
+            flags.append('--hard-links')
         if self.rsync_verbose:
             flags.append('--verbose')
         if self.rsync_dry:
             flags.append('-n')
+        if self.rsync_arbitopts:
+            for arbopt in self.rsync_arbitopts:
```
Comment (stdweird) [inline lib/vsc/zk/rsync/source.py]: make a temporary array with the flags, log all the arbitrary options with a `log.warning` and then do `flags.extend(arb_flags)`

File: lib/vsc/zk/rsync/source.py [hunk 4]
```diff
@@ -312,10 +314,15 @@ def get_flags(self, file, recursive):
             flags.append('--delete')
         if self.rsync_checksum:
             flags.append('--checksum')
+        if self.rsync_hardlinks:
+            flags.append('--hard-links')
         if self.rsync_verbose:
             flags.append('--verbose')
         if self.rsync_dry:
             flags.append('-n')
+        if self.rsync_arbitopts:
```
Comment (stdweird) [inline lib/vsc/zk/rsync/source.py]: add a comment that this should always be processed last (e.g. in case new ones are added)

**Comment by kwaegema** [comment]
remarks done

**Pull request merged**